csm/conf/uds.py

```diff
@@ -0,0 +1,164 @@
+# CORTX-CSM: CORTX Management web and CLI interface.
+# Copyright (c) 2020 Seagate Technology LLC and/or its Affiliates
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published
+# by the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+
+from csm.conf.salt import SaltWrappers
+from pathlib import Path
+from pwd import getpwnam
+from shutil import copyfile, rmtree
+from tempfile import NamedTemporaryFile
+
+import json
+import os
+
+
+class UDSConfigGenerator:
+    HAPROXY_BEGIN_UDS = '\n# BEGIN UDS\n'
+    HAPROXY_END_UDS = '\n# END UDS\n'
+    HAPROXY_UDS_WARNING = """\
+# The following HAproxy config entries, as well as the ``# BEGIN UDS`` and
+# ``# END UDS`` comment lines surrounding it, were automatically generated by
+# ``csm_setup``. Please *do not edit these manually*.
+# Only a single occurrence of an UDS-related config block is supported by
+# ``csm_setup``.
+"""
+    HAPROXY_CONFIG_PATH = '/etc/haproxy/haproxy.cfg'
+    HAPROXY_TEMP_CONFIG_PREFIX = 'haproxy.cfg.'
+    UDS_HOME_DIR = '/var/lib/uds'
+    UDS_CONFIG_DIR = f'{UDS_HOME_DIR}/.uds'
+    UDS_CONFIG_PATH = f'{UDS_CONFIG_DIR}/uds-config.json'
+    UDS_USERNAME = 'uds'
+
+    @staticmethod
+    def generate_haproxy_config():
+        cluster_ip = SaltWrappers.get_salt_call('pillar.get', 'cluster:cluster_ip')
+        minions = list(SaltWrappers.get_salt('grains.get', 'id').values())
+        # Codacy was complaining about the use of `assert` below :(
+        # assert len(minions) == 2
+        minions.sort()
+        return f"""\
+frontend uds-frontend
+    mode tcp
+    bind localhost4:5000
```

Comment:
what is localhost4 & localhost6 ?

Reply:
```
# grep localhost[46] /etc/hosts
127.0.0.1     localhost localhost.localdomain localhost4 localhost4.localdomain4
::1           localhost localhost.localdomain localhost6 localhost6.localdomain6
```

Reply:
Any reason why localhost was not used?
Use localhost and remove the second line with localhost6.

Reply:
`localhost` will resolve to either the IPv4 or the IPv6 address, so UDS would not support dual stack on the loopback interface. Using `localhost4` and `localhost6` ensures that support.

Reply:
By the way, if the problem is naming itself, I can replace those entries by `127.0.0.1` and `::1`.

If we wish to bind them to a specific stack such as IPv4, I can remove the IPv6 entry.

I just can't replace entries by `localhost` because that would depend on how name resolution works and might introduce unexpected behavior.

Reply:
It will be good to bind to 127.0.0.1.
I am not sure if localhost4 and localhost6 are standard names. I tried to find some documentation on internet but could not find any. Let me know if you have some references.

Reply:
They are not "standard", but neither is `localhost` itself (although it is certainly ubiquitous).

`localhost4` and `localhost6` are used by RedHat based distros and some others. I though it was OK to use them in that case.

Should I bind to `::1` as well? It will certainly make life easier and avoid nasty problems with `localhost` resolution.

Reply:
Fixed. Now using `127.0.0.1` and `::1` instead.